src/query/storages/common/index/src/filters/xor8/mod.rs

```diff
@@ -14,12 +14,135 @@
 
 mod block_filter;
 mod block_filter_versions;
+mod bloom_filter;
 mod xor8_filter;
 
+use std::hash::Hash;
+
 pub use block_filter::BlockFilter;
 pub use block_filter_versions::BlockBloomFilterIndexVersion;
 pub use block_filter_versions::V2BloomBlock;
+pub use bloom_filter::BloomBuilder;
+pub use bloom_filter::BloomFilter;
+use bytes::Bytes;
+use databend_common_exception::ErrorCode;
 pub use xor8_filter::Xor8Builder;
-pub use xor8_filter::Xor8BuildingError;
-pub use xor8_filter::Xor8CodecError;
 pub use xor8_filter::Xor8Filter;
+
+use crate::filters::Filter;
+use crate::filters::FilterBuilder;
+
+pub enum FilterImpl {
+    Xor(Xor8Filter),
+    Ngram(BloomFilter),
+}
+
+pub enum FilterImplBuilder {
+    Xor(Xor8Builder),
+    Ngram(BloomBuilder),
+}
+
+impl TryFrom<&FilterImpl> for Vec<u8> {
+    type Error = ErrorCode;
+
+    fn try_from(value: &FilterImpl) -> std::result::Result<Vec<u8>, ErrorCode> {
+        value.to_bytes()
+    }
+}
+
+impl TryFrom<Bytes> for FilterImpl {
+    type Error = ErrorCode;
+
+    fn try_from(value: Bytes) -> std::result::Result<FilterImpl, Self::Error> {
+        Ok(Self::from_bytes(value.as_ref())?.0)
+    }
+}
+
+impl FilterImpl {
+    pub fn mem_bytes(&self) -> usize {
+        match self {
+            FilterImpl::Xor(filter) => {
+                std::mem::size_of::<Xor8Builder>() + filter.filter.finger_prints.len()
+            }
+            FilterImpl::Ngram(_) => std::mem::size_of::<BloomFilter>(),
+        }
+    }
+}
+
+impl Filter for FilterImpl {
+    type CodecError = ErrorCode;
+
+    fn len(&self) -> Option<usize> {
+        match self {
+            FilterImpl::Xor(filter) => filter.len(),
+            FilterImpl::Ngram(filter) => filter.len(),
+        }
+    }
+
+    fn contains<K: ?Sized + Hash>(&self, key: &K) -> bool {
+        match self {
+            FilterImpl::Xor(filter) => filter.contains(key),
+            FilterImpl::Ngram(filter) => filter.contains(key),
+        }
+    }
+
+    fn contains_digest(&self, digest: u64) -> bool {
+        match self {
+            FilterImpl::Xor(filter) => filter.contains_digest(digest),
+            FilterImpl::Ngram(filter) => filter.contains_digest(digest),
+        }
+    }
+
+    fn to_bytes(&self) -> Result<Vec<u8>, Self::CodecError> {
+        Ok(match self {
+            FilterImpl::Xor(filter) => filter.to_bytes()?,
+            FilterImpl::Ngram(filter) => {
+                let mut bytes = filter.to_bytes()?;
+                bytes.splice(0..0, "bloom".as_bytes().iter().cloned());
+                bytes
+            }
+        })
+    }
+
+    fn from_bytes(buf: &[u8]) -> Result<(Self, usize), Self::CodecError> {
+        Ok(if buf.starts_with("bloom".as_bytes()) {
+            BloomFilter::from_bytes(&buf[5..])
```

Comment:
what if the old xor filter starts with `bloom` ?

I think we can read the first byte to indicate which kind of the filter.


Look at the `Cbor::decode` of `decode_hdr`

`major` ranges from [0, 7] is `Xor8Filter`

Let's check the first byte, if it's `n`, we can take it as Ngram, otherwise fallback to use `Xor`